### 1. 다음 코드의 실행 결과를 작성하세요.(민아) 
```java
class Main {
    public static void main(String[] args) {
        System.out.println(1) ;
        System.out.println(2);
        try {
            System.out.println(3);
            System.out.println(0/0);
            System.out.println(4) ;
        } catch (ArithmeticException ae) {
            System.out.println(5);
        }
        System.out.println (6);
    }
}
```

답)
<br>

```
1
2
3
5
6
```
1, 2, 3까지 출력한 후 7번째 line에서 0으로 나눠 ArithmeticException이라는 예외가 발생하여 try문을 빠져나와 catch문이 실행되어 5가 출력되고 마지막으로 6이 출력됩니다.

<br></br>

### 2. 다음에서 설명하는 메서드가 무엇인지 작성하세요.(민아)
a. 예외발생 당시의 호출스택(Call Stack)에 있었던 메서드의 정보와 예외 메시지를 화면에 출력한다.
<br>
b. 발생한 예외클래스의 인스턴스에 저장된 메시지를 얻을 수 있다.

<br></br>

### 3. 다음 코드의 실행 결과를 작성하세요.(민아)
```java
class Main {
    public static void main(String[] args) {
        int iVal = 100;
        String strVal = String. valueOf (iVal);

        double dVal = 200.0;
        String strVal2 = dVal + "";

        double sum = Integer.parseInt("+"+strVal)
        + Double.parseDouble(strVal2);

        double sum2 = Integer.valueOf(strVal) + Double.valueOf(strVal2);
        System.out.println(String.join("",strVal,"+", strVal2, "=")+sum);
        System.out.println(strVal+"+"+strVal2+"="+sum2) ;
    }
}
```
답변)
```
100+200.0=300.0
100+200.0=300.0
```
`String.valueOf` 는 인자값을 값으로 갖는 `String` 타입으로 반환<br>
`Integer.parseInt` 는 인자값을 정수형으로 변환<br>
`Double.parseDouble` 은 인자값을 실수형으로 변환<br>
`String.join` 은 문자열들을 이어 붙여 반환<br>
`+` 연산자는 `String` 타입에서 문자열을 이어 붙여 새로운 문자열을 반환

<br></br>

### 4. 예외처리의 정의와 목적을 설명하세요.(성훈)

<br></br>

### 5. 다음 코드의 실행 결과를 설명하세요.(성훈)

```java
public class Main {
    public static void main(String[] args) {
        try {
            method1();
        } catch (Exception e) {
            System.out.println("main메서드에서 예외가처리되었습니다.");
            e.printStackTrace();
        }
    }
    static void method1() throws Exception {
throw new Exception();
    }
}
```

<br>답변)
<br>

```
main메서드에서 예외가처리되었습니다.
java.lang.Exception
	at Main.method1(Main.java:11)
	at Main.main(Main.java:4)
```
try문의 method1()은 Exception 에러를 던지고 있다. <br>
따라서 Exception 에러를 catch하여 "main메서드에서 예외처리가 되었습니다." 라는 문구를 출력한다.<br>
printStackTrace() 메서드는 예외에 대한 정보를 출력한다.<br>
따라서 method1()인 11번째에서 오류가 발생하였고,<br> 
이 메서드를 호출한 main문 4번째 줄에서 오류가 발생하였다고 알려준다.<br>
</br>

### 6. 다음 코드의 실행 결과를 설명하세요.(성훈)

```java
public class Main {
    public static void main(String[] args) {
        Main.method1();
        System.out.println ("method1()의 수행을 마치고 main메서드로 돌아왔습니다.");
    }
    static void method1() {
        try {
            System.out.println ("method1()이 호출되었습니다. ") ;
            return;
        } catch (Exception e) {
            e.printStackTrace() ;
        } finally {
            System. out.println ("method1()의 finally블럭이 실행되었습니다.") ;
        }
    }
}
```

<br></br>

### 7. try-catch 문의 변형으로, 자동으로 자원을 반환할 때 사용하는 문은 무엇인가요? 그리고 해당 문을 사용하면 자동으로 호출되는 메서드와, 자동으로 호출되기 위해서는 클래스가 어떤 인터페이스를 구현해야 하는지 설명해주세요.(선민)

답)
try-with-resources 문
<br>
자동으로 호출되는 메서드는 close()이며,
AutoCloseable 인터페이스를 구현해야 합니다.
<br></br>

### 8. String 클래스의 불변성에 대해 설명하고, 문자열 결합이 빈번할 때 String 클래스 대신 어떤 클래스를 사용하면 좋은지, 해당 클래스들은 어떤 특징을 갖는지 설명해주세요.(선민)

답)
<br>
String 클래스는 immutable(불변) 불변성을 가졌기 때문에 String 객체를 변경시 실제로 변경되는 것이 아닌 새로운 String 객체를 생성합니다.
즉, 지정된 문자열을 변경할 수 없습니다.

StringBuffer<br>
문자열을 변경하거나 조작할 때 기존 객체의 내용을 바로 변경할 수 있다. 따라서 문자열 연산이 빈번하게 발생하는 상황에서 성능 향상을 얻을 수 있다.

StringBuilder<br>
가변성 (Mutability): StringBuilder 역시 StringBuffer처럼 문자열 조작 시 기존 객체의 내용을 변경한다. 따라서 문자열 연산이 빈번할 때 성능 향상을 얻을 수 있다.
동기화 없음: StringBuilder는 StringBuffer와 달리 동기화되어 있지 않다. 따라서 멀티 스레드 환경에서는 안전하지 않을 수 있지만, 단일 스레드 환경에서는 StringBuffer보다 더 빠른 성능을 제공한다.
StringBuilder는 StringBuffer와 매우 유사하며, 동일한 API를 대부분 가지고 있다.

차이점 요약 <br>
불변성 vs 가변성: String은 불변, StringBuffer와 StringBuilder는 가변
동기화: StringBuffer는 동기화되어 있어 멀티 스레드 환경에서 안전하나 성능에 불리할 수 있다. 반면, StringBuilder는 동기화되어 있지 않아 단일 스레드 환경에서 더 빠른 성능을 제공한다.
<br></br>

### 9. 사용자로부터 1 ~ 100 사이의 값을 입력받는 코드를 작성하세요. 만약 값이 해당 범위를 넘어설 경우 예외를 발생시키고, 예외 인스턴스를 통해 `1 ~ 100 사이의 값만 입력해주세요`라는 에러 메시지를 출력할 수 있도록 해주세요.(선민)

<br></br>

<br>

### 10. 아래의 코드에서 어떠한 예외도 던져지지 않은 경우 실행되는 코드는 무엇인지 설명하시오.(지영)

```java
try{ 
    // (1)
} catch(Exception1 e1){ 
    // (2) 
} catch(Exception2 e2){
    // (3) 
} finally{ 
    //(4) 
}
// (5)
```
답변) (1) -> (4) -> (5)<br>
finally 문은 예외 발생 여부와 무관하게 반드시 실행되는 문입니다.<br>
따라서 가장 먼저 try 문 내부 코드가 실행되고 여기서 아무런 예외도 발생하지 않으면 catch 문을 넘어 finally 문이 수행됩니다.<br>
그 이후에는 try - catch - finally 가 끝났으므로 추후 로직이 수행됩니다.

<br></br>

### 11. 예외처리문에서 catch 블록 없이 try-finally 블록만 가진 코드가 의도하는 것은 무엇인지 설명하고 사용 예시를 들어주세요.(지영)

```java
try{
    ...
} finally{
    ...
}

```
답)
try-finally블록은 try블록의 문장을 수행하는 동안에 예외의 발생여부에 관계없이 실행되어야 하는 코드를 포함시킬 목적으로 사용합니다.
<br>뿐만 아니라, 자원 관리, 부작용 방지, 코드 간결성 유지를 위해 활용될 수 있는 유용한 도구입니다.
<br>
<br>예시

```java
public class Main {
    public static void main(String[] args) {
        Main.method1();
        System.out.println("method1() 의 수행을 마치고 main메서드로 돌아왔습니다.");
    }
    static void method1() {
        try {
            System.out.println ("method1()이 호출되었습니다 . ") ;
            return;
        } finally {
            System.out.println ("method1()의 finally블럭이 실행되었습니다.");
        }
    }
}
```

<br></br>

### 12. 다음 코드의 출력 결과를 보고 equals() 메서드의 특징에 대해 설명하시오.(지영)

```java
public class Problem12{
    public static void main(String[] args) {
        String str1 = "abc";
        String str2 = new String("abc");
        String str3 = new String("abc");

        System.out.println("==연산자");
        System.out.println(str1 == "abc");
        System.out.println(str2 == "abc");

        System.out.println("equals 메서드");
        System.out.println(str1.equals("abc"));
        System.out.println(str2.equals("abc"));
        System.out.println(str2.equals(str3));
        
        System.out.println(str2.equals("ABC"));
    }
}
```

답)
<br>

```
==연산자
true
false
equals 메서드
true
true
true
false
```
`equals()` 메서드는 객체의 내용을 비교하는 데 사용되며, 두 객체의 값이 같으면 true를 반환합니다. 반면에 `==` 연산자는 객체의 참조를 비교하여 두 객체가 같은 메모리 주소를 가리키는지를 확인하며, 이는 두 객체가 동일한 객체인지를 확인합니다. 

<br></br>


### 13. 예외 되던지기가 무엇인지 설명하시오.(기훈)

<br></br>

### 14. java.lang 패키지의 특징과 자주사용되는 클래스 2가지를 설명하시오.(기훈)

<br>
`java.lang` 패키지는 자바프로그래밍에 가장 기본이 되는 클래스들을 포함하는 패키지입니다. <br>
이 패키지에 속한 클래스는 `import` 문 없이도 사용할 수 있다는 특징이 있습니다.<br>
대표적인 클래스로는 `String`, `System` 클래스 등이 있습니다.<br>
<br>

### 15. String클래스의 특징을 설명하시오. (기훈)

답)
1. String 클래스는 변경이 불가능한 클래스입니다. 
- 문자열 내용 변경이나 문자열에 문자 추가 또는 삭제가 불가능합니다.

2. 문자열 결합 연산자(+)를 지원합니다.

3. 다양한 문자열 조작을 위한 메서드를 제공합니다. 
- 이러한 메서드들을 사용하여 문자열을 검색, 추출, 분할, 연결 등의 작업을 수행할 수 있습니다.
- 예를 들어, length(), charAt(), substring(), concat(), indexOf(), split() 등의 메서드가 있습니다.

<br></br>

### 16. try-catch-finally 블록에 대한 설명 중 틀린 것은 무엇입니까? (현수)
1. try {} 블록에는 예외가 발생할 수 있는 코드를 작성한다.
2. catch {}블록은 try {} 블록에서 발생한 예외를 처리하는 블록이다.
3. try {} 블록에서 return 문을 사용하면 finally {} 블록은 실행되지 않는다.
4. catch {} 블록은 예외의 종류별로 여러 개를 작성할 수 있다.

<br></br>

### 17. throws에 대한 설명으로 틀린 것은 무엇입니까? (현수)
1. 생성자나 메소드의 선언 끝 부분에 사용되어 내부에서 발생된 예외를 떠넘긴다.
2. throws 뒤에는 떠넘겨야 할 예외를 쉼표(,)로 구분해서 기술한다.
3. 모든 예외를 더넘기기 위해 간단하게 throws Exception으로 작성할 수 있다.
4. 새로운 예외를 발생시키기 위해 사용된다.

답) 4번
<br>

throws는 메소드 선언부 끝에 작성되어 메소드에서 처리하지 않은 예외를 호출한 곳으로 떠넘기는 역할을 한다. 즉, 강제적으로 예외를 발생시켜주는 역할을 한다 ex) throw new Exception(); => Exception예외 발생

<br></br>

### 18. 메소드가 다음과 같이 선언되어 있습니다. 잘못된 예외 처리를 선택하세요. (현수)
```java
public void method1() throws NumberFormatException, ClassNotFoundException { ... }
```
1. try {method1():} catch (Exception e) {}
2. void method2() throws Exception {methodl(); }
3. try { method1(); }<br>
catch (Exception e) {}<br>
catch (ClassNotFoundException e) {}
4. try (method1(); }<br>
catch (ClassNotFoundException e) {}<br>
catch (NumberFormatException e) {}

답변) 3번<br>
try-catch 문은 예외가 발생하면 가장 위쪽의 catch 블럭 부터 발생한 예외를 처리할 수 있는지 확인하고 가장 먼저 처리할 수 있는 catch 블럭에서 처리하게 됩니다.<br>
3번 코드는 모든 예외 클래스의 부모 격인 `Exception` 클래스를 첫 번째 catch 블록에서 처리하도록 하였으므로 두 번째 catch 블록은 절대 도달할 수 없는 코드입니다.